emacs/config.org
#+TITLE: Levisa's Emacs Config
#+AUTHOR: Levisa Vanya
#+STARTUP: showeverything
#+OPTIONS: toc:2 num:nil ^:nil

* TABLE OF CONTENTS :toc:noexport:
- [[#program-to-load][PROGRAM TO LOAD]]
  - [[#elpaca-package-manager][Elpaca Package Manager]]
  - [[#load-evil-mode][Load Evil Mode]]
- [[#general-keybindings][General Keybindings]]
- [[#all-the-icons][ALL THE ICONS]]
- [[#buffer-move][BUFFER-MOVE]]
- [[#fonts][FONTS]]
  - [[#setting-the-font-face][Setting The Font Face]]
  - [[#zooming-inout][Zooming In/OuT]]
- [[#graphical-user-interface-and-feature-tweaks][GRAPHICAL USER INTERFACE AND FEATURE TWEAKS]]
  - [[#disable-startup-message][Disable Startup Message]]
  - [[#enable-visible-bell][Enable Visible Bell]]
  - [[#disable-menubar-toolbars-and-scroolbars][Disable Menubar, Toolbars and Scroolbars]]
  - [[#display-line-numbers-and-truncated-lines][Display Line Numbers and Truncated Lines]]
  - [[#tab-size][Tab Size]]
  - [[#make-esc-quit-prompts][Make ESC Quit Prompts]]
  - [[#disable-backup-file][Disable Backup File]]
  - [[#disable-auto-indent][Disable Auto Indent]]
  - [[#auto-closing-pairs][Auto-Closing Pairs]]
  - [[#disable-auto-save][Disable Auto-Save]]
  - [[#disable-column-number-mode][Disable Column Number Mode]]
  - [[#disable-line-number-mode][Disable Line Number Mode]]
- [[#lsp-support][LSP SUPPORT]]
  - [[#lsp-install][LSP INSTALL]]
  - [[#company-mode][COMPANY MODE]]
  - [[#minimal-setup][MINIMAL SETUP]]
  - [[#setup-lsp][SETUP LSP]]
- [[#org-mode][ORG MODE]]
  - [[#enabling-table-of-contents][Enabling Table of Contents]]
  - [[#enabling-org-bullets][Enabling Org Bullets]]
  - [[#disable-electric-indent][Disable Electric Indent]]
  - [[#source-code-block-tag-expansion][Source Code Block Tag Expansion]]
  - [[#diminish-org-indent-mode][Diminish Org Indent Mode]]
  - [[#org-level-headers][Org Level Headers]]
- [[#theme][THEME]]
  - [[#doom-themes][DOOM-THEMES]]
  - [[#doom-modeline][DOOM-MODELINE]]
- [[#which-key][WHICH-KEY]]
- [[#reload-emacs][RELOAD EMACS]]
- [[#diminish][DIMINISH]]
- [[#minibuffer-escape][MINIBUFFER ESCAPE]]
- [[#neotree][NEOTREE]]

* PROGRAM TO LOAD
** Elpaca Package Manager
#+BEGIN_SRC emacs-lisp
    (defvar elpaca-installer-version 0.6)
    (defvar elpaca-directory (expand-file-name "elpaca/" user-emacs-directory))
    (defvar elpaca-builds-directory (expand-file-name "builds/" elpaca-directory))
    (defvar elpaca-repos-directory (expand-file-name "repos/" elpaca-directory))
    (defvar elpaca-order '(elpaca :repo "https://github.com/progfolio/elpaca.git"
				  :ref nil
				  :files (:defaults "elpaca-test.el" (:exclude "extensions"))
				  :build (:not elpaca--activate-package)))
    (let* ((repo  (expand-file-name "elpaca/" elpaca-repos-directory))
	   (build (expand-file-name "elpaca/" elpaca-builds-directory))
	   (order (cdr elpaca-order))
	   (default-directory repo))
      (add-to-list 'load-path (if (file-exists-p build) build repo))
      (unless (file-exists-p repo)
	(make-directory repo t)
	(when (< emacs-major-version 28) (require 'subr-x))
	(condition-case-unless-debug err
	    (if-let ((buffer (pop-to-buffer-same-window "*elpaca-bootstrap*"))
		     ((zerop (call-process "git" nil buffer t "clone"
					   (plist-get order :repo) repo)))
		     ((zerop (call-process "git" nil buffer t "checkout"
					   (or (plist-get order :ref) "--"))))
		     (emacs (concat invocation-directory invocation-name))
		     ((zerop (call-process emacs nil buffer nil "-Q" "-L" "." "--batch"
					   "--eval" "(byte-recompile-directory \".\" 0 'force)")))
		     ((require 'elpaca))
		     ((elpaca-generate-autoloads "elpaca" repo)))
		(progn (message "%s" (buffer-string)) (kill-buffer buffer))
	      (error "%s" (with-current-buffer buffer (buffer-string))))
	  ((error) (warn "%s" err) (delete-directory repo 'recursive))))
      (unless (require 'elpaca-autoloads nil t)
	(require 'elpaca)
	(elpaca-generate-autoloads "elpaca" repo)
	(load "./elpaca-autoloads")))
    (add-hook 'after-init-hook #'elpaca-process-queues)
    (elpaca `(,@elpaca-order))
#+END_SRC

** Load Evil Mode
       
#+BEGIN_SRC emacs-lisp
  ;; Install use-package support
  (elpaca elpaca-use-package
  ;; Enable :elpaca use-package keyword.
  (elpaca-use-package-mode)
  ;; Assume :elpaca t unless otherwise specified.
  (setq elpaca-use-package-by-default t))

  ;; Block until current queue processed.
  (elpaca-wait)

  ;;When installing a package which modifies a form used at the top-level
  ;;(e.g. a package which adds a use-package key word),
  ;;use `elpaca-wait' to block until that package has been installed/configured.
  ;;For example:
  ;;(use-package general :demand t)
  ;;(elpaca-wait)

  ;; Expands to: (elpaca evil (use-package evil :demand t))
  (use-package evil
      :init
      (setq evil-want-integration t)
      (setq evil-want-keybinding nil)
      (setq evil-vsplit-window-right t)
      (setq evil-split-window-below t)
      (setq evil-want-C-u-scroll t)
      (evil-mode))
     (use-package evil-collection
       :after evil
       :config
       (setq evil-collection-mode-list '(dashboard dired ibuffer))
       (evil-collection-init))
     (use-package evil-tutor)
  ;;(with-eval-after-load 'evil-maps
  ;; (define-key evil-insert-state-map (kbd "TAB") 'tab-to-tab-stop))  
  ;;Turns off elpaca-use-package-mode current declaration
  ;;Note this will cause the declaration to be interpreted immediately (not deferred).
  ;;Useful for configuring built-in emacs features.
;; Using RETURN to follow links in Org/Evil 
;; Unmap keys in 'evil-maps if not done, (setq org-return-follows-link t) will not work
(with-eval-after-load 'evil-maps
  (define-key evil-motion-state-map (kbd "SPC") nil)
  (define-key evil-motion-state-map (kbd "RET") nil)
  (define-key evil-motion-state-map (kbd "TAB") nil))
;; Setting RETURN key in org-mode to follow links
  (setq org-return-follows-link  t)
  (use-package emacs :elpaca nil :config (setq ring-bell-function #'ignore))

  ;; Don't install anything. Defer execution of BODY
  ;;(elpaca nil (message "deferred"))
#+END_SRC
                  

* General Keybindings
#+BEGIN_SRC emacs-lisp
  (use-package general
   :config
   (general-evil-setup)
   ;; set up 'SPC' as the global leader key

   (general-create-definer Levisa/leader-keys
     :states '(normal insert visual emacs)
     :keymaps 'override
     :prefix "SPC" ;; set leader
     :global-prefix "M-SPC") ;; access leader insert mode

    (Levisa/leader-keys     
     "." '(find-file :wk "Find file")
     "f c" '((lambda () (interactive) (find-file "C:/Users/Moderator/.emacs.d/config.org")) :wk "Edit emacs config")
     "TAB TAB" '(comment-line :wk "Comment lines")
     "u" '(universal-argument :wk "Universal argument"))

    (Levisa/leader-keys
     "b" '(:ignore t :wk "buffer")
     "b b" '(switch-to-buffer :wk "Switch buffer")
     "b i" '(ibuffer :wk "Ibuffer")
     "b k" '(kill-this-buffer :wk "Kill this buffer")
     "b n" '(next-buffer :wk "Next buffer")
     "b p" '(previous-buffer :wk "Previous buffer")
     "b r" '(revert-buffer :wk "Reload buffer"))

    (Levisa/leader-keys
     "e" '(:ignore t :wk "Evaluate")    
     "e b" '(eval-buffer :wk "Evaluate elisp in buffer")
     "e d" '(eval-defun :wk "Evaluate defun containing or after point")
     "e e" '(eval-expression :wk "Evaluate and elisp expression")
     "e l" '(eval-last-sexp :wk "Evaluate elisp expression before point")
     "e r" '(eval-region :wk "Evaluate elisp in region")) 

    (Levisa/leader-keys
     "h" '(:ignore t :wk "Help")
     "h f" '(describe-function :wk "Describe function")
     "h v" '(describe-variable :wk "Describe variable")
     ;;"h r r" '((lambda () (interactive) (load-file "C:/Users/Moderator/.emacs.d/init.el")) :wk "Reload emacs config"))
     "h r r" '(reload-init-file :wk "Reload emacs config"))

    (Levisa/leader-keys
     "t" '(:ignore t :wk "Toggle")
     "t l" '(display-line-numbers-mode :wk "Toggle line numbers")
     "t t" '(visual-line-mode :wk "Toggle truncated lines")
     "t n" '(neotree-toggle :wk "Toggle neotree file viewer")
     "t o" '(org-mode :wk "Toggle org mode"))

    (Levisa/leader-keys
     "w" '(:ignore t :wk "Windows")
     ;; Window splits
     "w c" '(evil-window-delete :wk "Close window")
     "w n" '(evil-window-new :wk "New window")
     "w s" '(evil-window-split :wk "Horizontal split window")
     "w v" '(evil-window-vsplit :wk "Vertical split window"))

    ;; Window motions
    (Levisa/leader-keys
     "w h" '(evil-window-left :wk "Window left")
     "w j" '(evil-window-down :wk "Window down")
     "w k" '(evil-window-up :wk "Window up")
     "w l" '(evil-window-right :wk "Window right")
     "w w" '(evil-window-next :wk "Goto next window"))

    ;; Move Windows
    (Levisa/leader-keys
     "w H" '(buf-move-left :wk "Buffer move left")
     "w J" '(buf-move-down :wk "Buffer move down")
     "w K" '(buf-move-up :wk "Buffer move up")
     "w L" '(buf-move-right :wk "Buffer move right"))

  )
#+end_SRC
* ALL THE ICONS
#+begin_src emacs-lisp
(use-package all-the-icons
  :ensure t
  :if (display-graphic-p))

(use-package all-the-icons-dired
  :hook (dired-mode . (lambda () (all-the-icons-dired-mode t))))
#+end_src
* BUFFER-MOVE
#+begin_src emacs-lisp
(require 'windmove)

;;;###autoload
(defun buf-move-up ()
  "Swap the current buffer and the buffer above the split.
If there is no split, ie now window above the current one, an
error is signaled."
;;  "Switches between the current buffer, and the buffer above the
;;  split, if possible."
  (interactive)
  (let* ((other-win (windmove-find-other-window 'up))
	 (buf-this-buf (window-buffer (selected-window))))
    (if (null other-win)
        (error "No window above this one")
      ;; swap top with this one
      (set-window-buffer (selected-window) (window-buffer other-win))
      ;; move this one to top
      (set-window-buffer other-win buf-this-buf)
      (select-window other-win))))

;;;###autoload
(defun buf-move-down ()
"Swap the current buffer and the buffer under the split.
If there is no split, ie now window under the current one, an
error is signaled."
  (interactive)
  (let* ((other-win (windmove-find-other-window 'down))
	 (buf-this-buf (window-buffer (selected-window))))
    (if (or (null other-win) 
            (string-match "^ \\*Minibuf" (buffer-name (window-buffer other-win))))
        (error "No window under this one")
      ;; swap top with this one
      (set-window-buffer (selected-window) (window-buffer other-win))
      ;; move this one to top
      (set-window-buffer other-win buf-this-buf)
      (select-window other-win))))

;;;###autoload
(defun buf-move-left ()
"Swap the current buffer and the buffer on the left of the split.
If there is no split, ie now window on the left of the current
one, an error is signaled."
  (interactive)
  (let* ((other-win (windmove-find-other-window 'left))
	 (buf-this-buf (window-buffer (selected-window))))
    (if (null other-win)
        (error "No left split")
      ;; swap top with this one
      (set-window-buffer (selected-window) (window-buffer other-win))
      ;; move this one to top
      (set-window-buffer other-win buf-this-buf)
      (select-window other-win))))

;;;###autoload
(defun buf-move-right ()
"Swap the current buffer and the buffer on the right of the split.
If there is no split, ie now window on the right of the current
one, an error is signaled."
  (interactive)
  (let* ((other-win (windmove-find-other-window 'right))
	 (buf-this-buf (window-buffer (selected-window))))
    (if (null other-win)
        (error "No right split")
      ;; swap top with this one
      (set-window-buffer (selected-window) (window-buffer other-win))
      ;; move this one to top
      (set-window-buffer other-win buf-this-buf)
      (select-window other-win))))
#+end_src
* FONTS
** Setting The Font Face
#+BEGIN_SRC emacs-lisp
  (set-face-attribute 'default nil
    :font "CaskaydiaCove NF"
    :height 110
    :weight 'regular)
  (set-face-attribute 'fixed-pitch nil
    :font "CaskaydiaCove NF"
    :height 110
    :weight 'regular)
  (set-face-attribute 'font-lock-comment-face nil
    :slant 'italic)
  (set-face-attribute 'font-lock-keyword-face nil
    :slant 'italic)
  (add-to-list 'default-frame-alist '(font . "CaskaydiaCove NF-13"))
  (setq-default line-spacing 0.12)
#+END_SRC
** Zooming In/OuT
#+BEGIN_SRC emacs-lisp
(global-set-key (kbd "C-=") 'text-scale-increase)
(global-set-key (kbd "C--") 'text-scale-decrease)
(global-set-key (kbd "<C-wheel-up>") 'text-scale-increase)
(global-set-key (kbd "<C-wheel-down>") 'text-scale-decrease)
#+END_SRC
* GRAPHICAL USER INTERFACE AND FEATURE TWEAKS
** Disable Startup Message
#+begin_src emacs-lisp
(setq inhibit-startup-message t)
#+end_src

** Enable Visible Bell
#+begin_src emacs-lisp
(setq visible-bell t)
#+end_src

** Disable Menubar, Toolbars and Scroolbars
#+BEGIN_SRC emacs-lisp
(menu-bar-mode -1)
(tool-bar-mode -1)
(scroll-bar-mode -1)
#+END_SRC

** Display Line Numbers and Truncated Lines
#+BEGIN_SRC emacs-lisp
(global-display-line-numbers-mode 1)
(global-visual-line-mode t)
#+END_SRC

** Tab Size
#+BEGIN_SRC emacs-lisp
(setq-default tab-width 4)
#+END_SRC

** Make ESC Quit Prompts 
#+begin_src emacs-lisp
(global-set-key (kbd "<escape>") 'keyboard-escape-quit)
#+end_src

** Disable Backup File
#+begin_src emacs-lisp
(setq make-backup-files nil)
#+end_src

** Disable Auto Indent
#+begin_src emacs-lisp
(transient-mark-mode -1)
#+end_src

** Auto-Closing Pairs
#+begin_src emacs-lisp
(electric-pair-mode -1)
#+end_src

** Disable Auto-Save
#+begin_src emacs-lisp
(setq auto-save-default nil)
#+end_src>

** Disable Column Number Mode
#+begin_src emacs-lisp
(column-number-mode 0)
#+end_src

** Disable Line Number Mode
#+begin_src emacs-lisp
(line-number-mode 0)
#+end_src
* LSP SUPPORT
** LSP INSTALL
#+begin_src emacs-lisp
  ;; (use-package lsp-mode
  ;;  :ensure t
  ;;  :commands (lsp lsp-deferred)
  ;;  :init)
  ;; (use-package lsp-treemacs
  ;;  :ensure t) 
  ;;  (with-eval-after-load 'lsp-mode
  ;;  (add-hook 'lsp-mode-hook #'lsp-enable-which-key-integration))
#+end_src
** COMPANY MODE
#+begin_src emacs-lisp
  ;;   (use-package company
  ;;     :defer 2
  ;;     :diminish
  ;;     :after lsp-mode
  ;;     :hook (lsp-mode . company-mode)
  ;;     :bind (:map company-active-map
  ;;            ("<tab>" . company-complete-selection))
  ;;           (:map lsp-mode-map
  ;;            ("<tab>" . company-indent-or-complete-common))
  ;;     :custom
  ;;     (company-begin-commands '(self-insert-command))
  ;;     (company-idle-delay .1)
  ;;     (company-minimum-prefix-length 2)
  ;;     (company-show-numbers t)
  ;;     (company-tooltip-align-annotations 't)
  ;;     (global-company-mode t))

  ;;   (use-package company-box
  ;;     :after company
  ;;     :diminish
  ;;     :hook (company-mode . company-box-mode))
  ;; (setq company-tooltip-align-annotations t)
  ;; ;(setq company-vscode-light-icons-margin t)
  ;; (setq company-format-margin-function #'company-vscode-light-icons-margin)
#+end_src
** MINIMAL SETUP
#+begin_src emacs-lisp
  ;; (setq gc-cons-threshold 100000000)
  ;; (setq lsp-completion-provider :capf)
  ;; (setq lsp-idle-delay 0.500)
  ;; (setq lsp-log-io nil)
  ;; (setq lsp-enable-links nil)
  ;; (setq lsp-signature-render-documentation nil)
  ;; (setq lsp-headerline-breadcrumb-enable nil)
  ;; (setq lsp-ui-doc-enable nil)
  ;; (setq lsp-completion-enable-additional-text-edit nil)
#+end_src
** SETUP LSP 
*** Python
#+begin_src emacs-lisp
  ;; (use-package python-mode
  ;;  :mode "\\.py\\'"
  ;;  :hook (python-mode . lsp-deferred)
  ;;  :config
  ;; (setq python-indent-level 2))
#+end_src
*** Disable Python Warning
#+begin_src emacs-lisp
;;(setq lsp-pyls-plugins-pycodestyle-enabled nil)
#+end_src
*** LSP Features
#+begin_src emacs-lisp
;;(setq lsp-ui-sideline-enable nil)
;;(setq lsp-modeline-code-actions-enable nil)
;;(setq lsp-ui-sideline-enable t)
;;(setq lsp-modeline-diagnostics-enable nil)
;;(setq lsp-ui-sideline-enable t)
;;(setq lsp-signature-auto-activate t)
;;(setq lsp-signature-render-documentation nil)

#+end_src
* ORG MODE
** Enabling Table of Contents
#+BEGIN_SRC emacs-lisp
(use-package toc-org
 :commands toc-org-enable
 :init (add-hook 'org-mode-hook 'toc-org-enable))
#+END_SRC

** Enabling Org Bullets
#+BEGIN_SRC emacs-lisp
(add-hook 'org-mode-hook 'org-indent-mode)
(use-package org-bullets)
(add-hook 'org-mode-hook (lambda () (org-bullets-mode 1)))
#+END_SRC

** Disable Electric Indent
#+BEGIN_SRC emacs-lisp
(electric-indent-mode -1)
#+END_SRC

** Source Code Block Tag Expansion
| Typing the below + TAB | Expands to ...                          |
|------------------------+-----------------------------------------|
| <a                     | '#+BEGIN__EXPORT ascii' … '#+END_EXPORT |
| <c                     | '#+BEGIN__CENTER' … '#+END_CENTER'      |
| <C                     | '#+BEGIN__COMMENT' … '#+END_COMMENT'    |
| <e                     | '#+BEGIN__EXAMPLE' … '#+END_EXAMPLE'    |
| <E                     | '#+BEGIN__EXPORT' … '#+END_EXPORT'      |
| <h                     | '#+BEGIN__EXPORT html' … '#+END_EXPORT' |
| <l                     | '#+BEGIN__EXPORT latex' … '#+END_EXPORT' |
| <q                     | '#+BEGIN__QUOTE' … '#+END_QUOTE'        |
| <s                     | '#+BEGIN__SRC' … '#+END_SRC'            |
| <v                     | '#+BEGIN__VERSE' … '#+END_VERSE'        |
#+BEGIN_SRC emacs-lisp
(require 'org-tempo) 
#+END_SRC

** Diminish Org Indent Mode
#+begin_src emacs-lisp
(eval-after-load 'org-indent '(diminish 'org-indent-mode))
#+end_src

** Org Level Headers
#+begin_src emacs-lisp
(custom-set-faces
 '(org-level-1 ((t (:inherit outline-1 :height 1.7))))
 '(org-level-2 ((t (:inherit outline-2 :height 1.6))))
 '(org-level-3 ((t (:inherit outline-3 :height 1.5))))
 '(org-level-4 ((t (:inherit outline-4 :height 1.4))))
 '(org-level-5 ((t (:inherit outline-5 :height 1.3))))
 '(org-level-6 ((t (:inherit outline-5 :height 1.2))))
 '(org-level-7 ((t (:inherit outline-5 :height 1.1)))r 
#+end_src
* THEME			
#+begin_src emacs-lisp
;;(add-to-list 'custom-theme-load-path "C:/Users/Moderator/.emacs.d/themes")
;;(load-theme 'kanagawa t)
#+end_src

** DOOM-THEMES
#+begin_src emacs-lisp
(use-package doom-themes
  :ensure t
  :config
  ;; Global settings (defaults)
  (setq doom-themes-enable-bold t    ; if nil, bold is universally disabled
        doom-themes-enable-italic t) ; if nil, italics is universally disabled
  (load-theme 'doom-tokyo-night t) 
  ;; Enable flashing mode-line on errors
  (doom-themes-visual-bell-config)
  ;; Enable custom neotree theme (all-the-icons must be installed!)
  (doom-themes-neotree-config)
  ;; or for treemacs users
  (setq doom-themes-treemacs-theme "doom-atom") ; use "doom-colors" for less minimal icon theme
  (doom-themes-treemacs-config)
  ;; Corrects (and improves) org-mode's native fontification.
  (doom-themes-org-config))
#+end_src

** DOOM-MODELINE
#+begin_src emacs-lisp
  (use-package doom-modeline
    :ensure t
    :init   (doom-modeline-mode 1))
  (setq doom-modeline-height 25)
  (setq doom-modeline-bar-width 4)
  (setq doom-modeline-hud nil) 
  (setq doom-modeline-project-detection 'auto)
  (setq doom-modeline-buffer-file-name-style 'auto)

  ;; Whether display icons in the mode-line.
  ;; While using the server mode in GUI, should set the value explicitly.
  ;; (setq doom-modeline-icon t)

  ;; ;; Whether display the icon for `major-mode'. It respects option `doom-modeline-icon'.
  ;; (setq doom-modeline-major-mode-icon t)

  ;; ;; Whether display the colorful icon for `major-mode'.
  ;; ;; It respects `nerd-icons-color-icons'.
  ;; (setq doom-modeline-major-mode-color-icon t)

  ;; ;; Whether display the icon for the buffer state. It respects option `doom-modeline-icon'.
  ;; (setq doom-modeline-buffer-state-icon t)

  ;; ;; Whether display the modification icon for the buffer.
  ;; ;; It respects option `doom-modeline-icon' and option `doom-modeline-buffer-state-icon'.
  ;; (setq doom-modeline-buffer-modification-icon t)

  ;; ;; Whether display the lsp icon. It respects option `doom-modeline-icon'.
  ;; (setq doom-modeline-lsp-icon t)

  ;; ;; Whether display the time icon. It respects option `doom-modeline-icon'.
  ;; (setq doom-modeline-time-icon t)

  ;; ;; Whether display the live icons of time.
  ;; ;; It respects option `doom-modeline-icon' and option `doom-modeline-time-icon'.
  ;; (setq doom-modeline-time-live-icon t)

  ;; ;; Whether to use unicode as a fallback (instead of ASCII) when not using icons.
  ;; (setq doom-modeline-unicode-fallback nil)

  ;; ;; Whether display the buffer name.
  ;; (setq doom-modeline-buffer-name t)

  ;; ;; Whether highlight the modified buffer name.
  ;; (setq doom-modeline-highlight-modified-buffer-name t)

  ;; ;; When non-nil, mode line displays column numbers zero-based.
  ;; ;; See `column-number-indicator-zero-based'.
  ;; (setq doom-modeline-column-zero-based nil)

  ;; ;; Specification of \"percentage offset\" of window through buffer.
  ;; ;; See `mode-line-percent-position'.
  ;; ;;(setq doom-modeline-percent-position '(-3 "%p"))
   (setq doom-modeline-percent-position nil)
  ;; ;; Format used to display line numbers in the mode line.
  ;; ;; See `mode-line-position-line-format'.
  ;; ;; (setq doom-modeline-position-line-format '("L%l"))
  ;; (setq doom-modeline-position-line-format '())
  ;; ;; Format used to display column numbers in the mode line.
  ;; ;; See `mode-line-position-column-format'.
  ;; ;; (setq doom-modeline-position-column-format '("C%c"))
  ;; (setq doom-modeline-position-column-format '())
  ;; ;; Format used to display combined line/column numbers in the mode line. See `mode-line-position-column-line-format'.
  ;; ;; (setq doom-modeline-position-column-line-format '("%l:%c"))
  ;; (setq doom-modeline-position-column-line-format '())
  ;; ;; Whether display the minor modes in the mode-line.
  ;; (setq doom-modeline-minor-modes nil)

  ;; ;; If non-nil, a word count will be added to the selection-info modeline segment.
  ;; (setq doom-modeline-enable-word-count nil)

  ;; ;; Major modes in which to display word count continuously.
  ;; ;; Also applies to any derived modes. Respects `doom-modeline-enable-word-count'.
  ;; ;; If it brings the sluggish issue, disable `doom-modeline-enable-word-count' or
  ;; ;; remove the modes from `doom-modeline-continuous-word-count-modes'.
  ;; (setq doom-modeline-continuous-word-count-modes '(markdown-mode gfm-mode org-mode))

  ;; ;; Whether display the buffer encoding.
  (setq doom-modeline-buffer-encoding nil)

  ;; ;; Whether display the indentation information.
  ;; (setq doom-modeline-indent-info nil)

  ;; ;; Whether display the total line number。
  ;; (setq doom-modeline-total-line-number nil)

  ;; ;; If non-nil, only display one number for checker information if applicable.
  ;; (setq doom-modeline-checker-simple-format t)

  ;; ;; The maximum number displayed for notifications.
  ;; (setq doom-modeline-number-limit 99)

  ;; ;; The maximum displayed length of the branch name of version control.
  ;; (setq doom-modeline-vcs-max-length 12)

  ;; ;; Whether display the workspace name. Non-nil to display in the mode-line.
  ;; (setq doom-modeline-workspace-name t)

  ;; ;; Whether display the perspective name. Non-nil to display in the mode-line.
  ;; (setq doom-modeline-persp-name t)

  ;; ;; If non nil the default perspective name is displayed in the mode-line.
  ;; (setq doom-modeline-display-default-persp-name nil)

  ;; ;; If non nil the perspective name is displayed alongside a folder icon.
  ;; (setq doom-modeline-persp-icon t)

  ;; ;; Whether display the `lsp' state. Non-nil to display in the mode-line.
  ;; (setq doom-modeline-lsp nil)

  ;; ;; Whether display the GitHub notifications. It requires `ghub' package.
  ;; (setq doom-modeline-github nil)

  ;; ;; The interval of checking GitHub.
  ;; (setq doom-modeline-github-interval (* 30 60))

  ;; ;; Whether display the modal state.
  ;; ;; Including `evil', `overwrite', `god', `ryo' and `xah-fly-keys', etc.
  ;; (setq doom-modeline-modal t)

  ;; ;; Whether display the modal state icon.
  ;; ;; Including `evil', `overwrite', `god', `ryo' and `xah-fly-keys', etc.
  ;; (setq doom-modeline-modal-icon t)

  ;; ;; Whether display the modern icons for modals.
  ;; (setq doom-modeline-modal-modern-icon t)

  ;; ;; When non-nil, always show the register name when recording an evil macro.
  ;; (setq doom-modeline-always-show-macro-register nil)

  ;; ;; Whether display the gnus notifications.
  ;; (setq doom-modeline-gnus nil)

  ;; ;; Whether gnus should automatically be updated and how often (set to 0 or smaller than 0 to disable)
  ;; (setq doom-modeline-gnus-timer 2)

  ;; ;; Wheter groups should be excludede when gnus automatically being updated.
  ;; (setq doom-modeline-gnus-excluded-groups '("dummy.group"))

  ;; ;; Whether display the IRC notifications. It requires `circe' or `erc' package.
  ;; (setq doom-modeline-irc t)

  ;; ;; Function to stylize the irc buffer names.
  ;; (setq doom-modeline-irc-stylize 'identity)

  ;; ;; Whether display the battery status. It respects `display-battery-mode'.
   (setq doom-modeline-battery t)
   (display-battery-mode 1)
  ;; ;; Whether display the time. It respects `display-time-mode'.
   (setq doom-modeline-time t)
   (display-time-mode 1)
  ;; ;; Whether display the misc segment on all mode lines.
  ;; ;; If nil, display only if the mode line is active.
  ;; (setq doom-modeline-display-misc-in-all-mode-lines t)

  ;; ;; The function to handle `buffer-file-name'.
  ;; (setq doom-modeline-buffer-file-name-function #'identity)

  ;; ;; The function to handle `buffer-file-truename'.
  ;; (setq doom-modeline-buffer-file-truename-function #'identity)

  ;; ;; Whether display the environment version.
  ;; (setq doom-modeline-env-version t)
  ;; ;; Or for individual languages
  ;; (setq doom-modeline-env-enable-python t)

emacs/config.el
(defvar elpaca-installer-version 0.6)
  (defvar elpaca-directory (expand-file-name "elpaca/" user-emacs-directory))
  (defvar elpaca-builds-directory (expand-file-name "builds/" elpaca-directory))
  (defvar elpaca-repos-directory (expand-file-name "repos/" elpaca-directory))
  (defvar elpaca-order '(elpaca :repo "https://github.com/progfolio/elpaca.git"
				:ref nil
				:files (:defaults "elpaca-test.el" (:exclude "extensions"))
				:build (:not elpaca--activate-package)))
  (let* ((repo  (expand-file-name "elpaca/" elpaca-repos-directory))
	 (build (expand-file-name "elpaca/" elpaca-builds-directory))
	 (order (cdr elpaca-order))
	 (default-directory repo))
    (add-to-list 'load-path (if (file-exists-p build) build repo))
    (unless (file-exists-p repo)
      (make-directory repo t)
      (when (< emacs-major-version 28) (require 'subr-x))
      (condition-case-unless-debug err
	  (if-let ((buffer (pop-to-buffer-same-window "*elpaca-bootstrap*"))
		   ((zerop (call-process "git" nil buffer t "clone"
					 (plist-get order :repo) repo)))
		   ((zerop (call-process "git" nil buffer t "checkout"
					 (or (plist-get order :ref) "--"))))
		   (emacs (concat invocation-directory invocation-name))
		   ((zerop (call-process emacs nil buffer nil "-Q" "-L" "." "--batch"
					 "--eval" "(byte-recompile-directory \".\" 0 'force)")))
		   ((require 'elpaca))
		   ((elpaca-generate-autoloads "elpaca" repo)))
	      (progn (message "%s" (buffer-string)) (kill-buffer buffer))
	    (error "%s" (with-current-buffer buffer (buffer-string))))
	((error) (warn "%s" err) (delete-directory repo 'recursive))))
    (unless (require 'elpaca-autoloads nil t)
      (require 'elpaca)
      (elpaca-generate-autoloads "elpaca" repo)
      (load "./elpaca-autoloads")))
  (add-hook 'after-init-hook #'elpaca-process-queues)
  (elpaca `(,@elpaca-order))
;; Install a package via the elpaca macro
;; See the "recipes" section of the manual for more details.
;; (elpaca example-package)

;; Install use-package support
(elpaca elpaca-use-package
;; Enable :elpaca use-package keyword.
(elpaca-use-package-mode)
;; Assume :elpaca t unless otherwise specified.
(setq elpaca-use-package-by-default t))

;; Block until current queue processed.
(elpaca-wait)

;;When installing a package which modifies a form used at the top-level
;;(e.g. a package which adds a use-package key word),
;;use `elpaca-wait' to block until that package has been installed/configured.
;;For example:
;;(use-package general :demand t)
;;(elpaca-wait)

;; Expands to: (elpaca evil (use-package evil :demand t))
(use-package evil
	:init
    (setq evil-want-integration t)
    (setq evil-want-keybinding nil)
    (setq evil-vsplit-window-right t)
    (setq evil-split-window-below t)
    (evil-mode))
   (use-package evil-collection
     :after evil
     :config
     (setq evil-collection-mode-list '(dashboard dired ibuffer))
     (evil-collection-init))
   (use-package evil-tutor)

;;Turns off elpaca-use-package-mode current declaration
;;Note this will cause the declaration to be interpreted immediately (not deferred).
;;Useful for configuring built-in emacs features.
(use-package emacs :elpaca nil :config (setq ring-bell-function #'ignore))

;; Don't install anything. Defer execution of BODY
(elpaca nil (message "deferred"))

(use-package general
 :config
 (general-evil-setup)
 ;; set up 'SPC' as the global leader key

 (general-create-definer Levisa/leader-keys
   :states '(normal insert visual emacs)
   :keymaps 'override
   :prefix "SPC" ;; set leader
   :global-prefix "M-SPC") ;; access leader insert mode

 (Levisa/leader-keys
   "b" '(:ignore t :wk "buffer")
   "bb" '(switch-to-buffer :wk "Switch buffer")
   "bk" '(kill-this-buffer :wk "Kill this buffer")
   "bn" '(next-buffer :wk "Next buffer")
   "bp" '(previous-buffer :wk "Previous buffer")
   "br" '(revert-buffer :wk "Reload buffer"))
)

(set-face-attribute 'default nil
  :font "CaskaydiaCove NF"
  :height 110
  :weight 'regular)
(set-face-attribute 'fixed-pitch nil
  :font "CaskaydiaCove NF"
  :height 110
  :weight 'regular)
(set-face-attribute 'font-lock-comment-face nil
  :slant 'italic)
(set-face-attribute 'font-lock-keyword-face nil
  :slant 'italic)
(add-to-list 'default-frame-alist '(font . "CaskaydiaCove NF-13"))
(setq-default line-spacing 0.12)

(menu-bar-mode -1)
(tool-bar-mode -1)
(scroll-bar-mode -1)

(global-display-line-numbers-mode 1)
(global-visual-line-mode t)

(use-package which-key
	:init
		(which-key-mode 1)
	:config
		(setq which-key-side-window-location 'bottom
			  which-key-sort-order #'which-key-key-order-alpha
			  which-key-sort-uppercase-first nil
			  which-key-add-column-padding 1
			  which-key-max-display-columns nil
			  which-key-min-display-lines 6
			  which-key-side-window-slot -10
			  which-key-side-window-max-height 0.25
			  which-key-idle-delay 0.8
			  which-key-max-description-length 25
			  which-key-allow-imprecise-window-fit t
              which-key-separator "  "))
